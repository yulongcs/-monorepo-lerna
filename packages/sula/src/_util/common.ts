import assignWith from 'lodash/assignWith';
import isUndefined from 'lodash/isUndefined';
import isArray from 'lodash/isArray';

function withDefined(dest: any, src: any) {
  return isUndefined(src) ? dest : src;
}

export function assignWithDefined(...args: Record<string, any>[]) {
  return assignWith(...args, withDefined);
}

export function toArray<T>(value?: T | T[] | null): T[] {
  if (value === undefined || value === null) {
    return [];
  }

  // @ts-ignore
  return isArray(value) ? value : [value];
}


