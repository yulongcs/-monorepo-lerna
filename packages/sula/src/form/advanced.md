---
title: 高级
---

<code title="advanced" src="./demos/advanced.tsx" />