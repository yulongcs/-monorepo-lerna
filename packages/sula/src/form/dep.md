---
title: Form依赖
---

<code title="dep" src="./demos/dep.tsx" />

<code title="match any" src="./demos/anymatch.tsx" />

<code title="group dep" src="./demos/groupdep.tsx" />