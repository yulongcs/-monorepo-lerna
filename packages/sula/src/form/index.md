---
title: Form
---

<code title="basic" src="./demos/basic.tsx" />

<code title="component" src="./demos/component.tsx" />

<code title="api" src="./demos/api.tsx" />

<code title="formaction" src="./demos/formaction.tsx" />


<code title="mediaqueries" src="./demos/mediaqueries.tsx" />