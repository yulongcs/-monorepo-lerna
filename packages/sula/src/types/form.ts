export type Mode = 'create' | 'edit' | 'view';

export type FieldNameList = string[];
export type FieldNamePath = string | FieldNameList;

export type FieldSource = Array<{
  text: string;
  value: string | number | boolean;
  [key: string]: any;
}>;
