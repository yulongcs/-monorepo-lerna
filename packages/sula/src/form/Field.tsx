import React from 'react';
import isUndefined from 'lodash/isUndefined';
import uniqueId from 'lodash/uniqueId';
import omit from 'lodash/omit';
import assign from 'lodash/assign';
import { FormItemProps } from 'antd/lib/form/FormItem';
import { Rule } from 'antd/lib/form';
import { FieldPlugin, ValidatorPlugin, RenderPlugin } from '../types/plugin';
import { FieldNamePath, FieldNameList, FieldSource } from '../types/form';
import { RequestConfig } from '../types/request';
import { Dependencies } from '../types/dependency';
import { toArray, assignWithDefined } from '../_util/common';
import FieldGroupContext, { HOOK_MARK } from './FieldGroupContext';
import {
  triggerFieldPlugin,
  triggerActionPlugin,
  triggerRenderPlugin,
  triggerPlugin,
} from '../rope/triggerPlugin';
import { Form as AForm, Col } from 'antd';
import { needWrapCols } from './utils/layoutUtil';
import FormDependency from './dependency';
import { ItemLayout, Layout } from './FieldGroup';
import { matchNameList } from '../_util/NameListMap';

const FormItem = AForm.Item;

export interface FieldProps
  extends Omit<FormItemProps, 'children' | 'wrapperCol' | 'labelCol' | 'rules'> {
  field: FieldPlugin;
  name?: FieldNamePath;
  collect?: boolean;
  initialDisabled?: boolean;
  initialVisible?: boolean;
  initialSource?: FieldSource;
  initialValue?: any;
  remoteSource?: RequestConfig;
  dependency?: Dependencies;
  children?: React.ReactElement;
  itemLayout?: ItemLayout;
  layout?: Layout;
  rules?: Array<
    Rule & {
      validator?: ValidatorPlugin;
    }
  >;
  childrenContainer?: RenderPlugin;
}

export default class Field extends React.Component<FieldProps> {
  static contextType = FieldGroupContext;

  private inited: boolean = false;

  private destroy = false;

  /**
   * 以下三个值，是针对field-form的扩展，主要是应用于render插件中的。
   */
  private source: FieldSource;

  private disabled: boolean;

  // 收集表单值相关
  private visible: boolean;

  private collect: boolean;

  // 如果不设置则给一个唯一值，并不会作为 FormItem 的 name
  private fieldName: string;

  private cancelRegisterField: () => void | null = null;

  private prevName: FieldNamePath;

  public mountedAndNeverUpdate: boolean = false;

  componentDidMount() {
    this.mountedAndNeverUpdate = true;

    if (!this.props.name) {
      return;
    }
    const {
      registerField,
      getFormDependency,
      getCtx,
      linkFieldNameAndFieldKey,
    } = this.context.formContext.getInternalHooks(HOOK_MARK);

    this.initFieldSource(getCtx());

    const { parentGroupName } = this.context;
    this.cancelRegisterField = registerField(parentGroupName, this);
    linkFieldNameAndFieldKey(this.getName(), this.getName(true));

    if (!this.props.dependency) {
      return;
    }

    // field 依赖
    const formDependency: FormDependency = getFormDependency();

    formDependency.parseFormDependency(this.props, this.getFieldNameList);
  }

  componentDidUpdate(prevProps: FieldProps) {
    this.mountedAndNeverUpdate = false;
    if (!this.props.name) {
      return;
    }

    /**
     * fieldKey 代表是动态表单项，如果更新前后两次name不一样，说明只是在values的位置上发生改变（上面增加了，或者删除了）
     */
    if (this.props.fieldKey && !matchNameList(toArray(this.props.name), toArray(prevProps.name))) {
      this.prevName = prevProps.name!;
      const {
        linkFieldNameAndFieldKey,
        unlinkFieldNameAndFieldKey,
      } = this.context.formContext.getInternalHooks(HOOK_MARK);

      // 动态添加
      unlinkFieldNameAndFieldKey(this.getName());
      linkFieldNameAndFieldKey(this.getName(), this.getName(true));
    }
  }

  componentWillUnmount() {
    if (!this.props.name) {
      return;
    }
    const {
      getFormDependency,
      unlinkFieldNameAndFieldKey,
    } = this.context.formContext.getInternalHooks(HOOK_MARK);

    // 删除依赖信息（主动）
    const formDependency: FormDependency = getFormDependency();
    formDependency.removeDependency(this.getName());

    // 动态删除
    unlinkFieldNameAndFieldKey(this.getName());

    this.cancelRegister();
    this.destroy = true;
  }

  private cancelRegister = () => {
    if (this.cancelRegisterField) {
      this.cancelRegisterField();
    }
    this.cancelRegisterField = null;
  };

  private initFieldSource = (ctx) => {
    if (this.props.remoteSource && this.props.remoteSource.init !== false) {
      triggerActionPlugin(ctx, this.props.remoteSource).then((data: any) => {
        /**
         * 如果有fieldKey，则使用fieldKey注册
         */
        ctx.form.setFieldSource(this.getName(), data);
      });
    }
  };

  public getFieldNameList = (name: FieldNamePath) => {
    const { isList, parentGroupName } = this.context;

    if (isList) {
      /**
       * name 是加了rc formList 数字的
       * paranGroupName 是 FormList的name
       */
      return getFieldName(parentGroupName, name);
    }
    return toArray(name);
  };

  public getName = (needFieldNameList?: boolean): FieldNameList | undefined => {
    /** fieldKey 只在动态增减使用，此时fieldKey是确定的，name是变化的 */
    const { name, fieldKey } = this.props;
    const finalName = (!isUndefined(fieldKey) && !needFieldNameList
      ? fieldKey
      : name) as FieldNamePath;
    if (isUndefined(finalName)) {
      return;
    }

    return this.getFieldNameList(finalName);
  };

  public getPrevName = () => {
    return this.prevName ? this.getFieldNameList(this.prevName) : this.getName(true);
  };

  public getSource() {
    return this.source;
  }

  public getDisabled() {
    return this.disabled;
  }

  public setSource = (source: any) => {
    this.source = source;
    this.reRender();
  };

  public setVisible = (visible: boolean) => {
    this.visible = visible;
    this.reRender();
  };

  public setDisabled = (disabled: boolean) => {
    this.disabled = disabled;
    this.reRender();
  };

  public getVisible() {
    return this.visible;
  }

  public getCollect() {
    return this.collect;
  }

  public reRender() {
    if (this.destroy) return;
    this.forceUpdate();
  }

  /**
   *
   */
  private transChildrenToElems = (ctx, children) => {
    const childrenElems = [];
    // 配置方式全部转换为组件
    children.forEach((childConfig, index) => {
      const childKey = childConfig.name || `${this.fieldName}-child-${index}`;
      let childElem;
      if (childConfig.field) {
        childElem = <Field {...childConfig} key={childKey} />;
      } else if (childConfig.render) {
        // 认为是render
        const renderPlugin = childConfig.render;
        const finalChildConfig = omit(childConfig, ['render']);
        childElem = (
          <Field key={childKey} {...finalChildConfig}>
            {triggerRenderPlugin(ctx, renderPlugin)}
          </Field>
        );
      }

      childrenElems.push(childElem);
    });

    return childrenElems;
  };

  private renderField(ctx, fieldConfig: FieldPlugin, extraConf) {
    const { itemLayout, visible, childrenContainer, formItemProps, isList } = extraConf;

    const { children, valuePropName = 'value', noStyle } = formItemProps;

    /** 如果是 isList 可能没有 itemLayout */
    const { wrapperCol, labelCol } = itemLayout;

    let fieldElem;

    if (children) {
      if (React.isValidElement(children)) {
        fieldElem = children;
      } else {
        // 配置型children
        fieldElem = this.transChildrenToElems(ctx, children);
        if (childrenContainer) {
          const childrenContainerElem = triggerRenderPlugin(ctx, childrenContainer);
          fieldElem = React.cloneElement(childrenContainerElem as React.ReactElement, {
            children: fieldElem,
          });
        }
      }
    } else {
      fieldElem = triggerFieldPlugin(ctx, fieldConfig, valuePropName);
    }

    const fieldItemElem = (
      <FormItem
        labelCol={isList ? labelCol || { span: 0 } : labelCol}
        wrapperCol={isList ? wrapperCol || { span: 24 } : wrapperCol}
        {...formItemProps}
      >
        {fieldElem}
      </FormItem>
    );

    if (noStyle !== true && !isList && (needWrapCols(itemLayout.span) || itemLayout.offset)) {
      return (
        <Col
          style={{ display: visible === false ? 'none' : '' }}
          span={itemLayout.span}
          offset={itemLayout.offset}
        >
          {fieldItemElem}
        </Col>
      );
    } else {
      const display = formItemProps.style ? formItemProps.style.display : '';
      return React.cloneElement(fieldItemElem, {
        style: assign({}, fieldItemElem.props.style, {
          display: visible === false ? 'none' : display,
        }),
      });
    }
  }

  render() {
    const {
      field: fieldProps,
      initialSource,
      initialVisible = true,
      initialDisabled = false,
      collect,
      remoteSource,
      itemLayout, // 无cols
      childrenContainer,
      ...restProps
    } = this.props;

    const { formContext, layout, isList, parentGroupName } = this.context;

    if (!this.inited) {
      this.inited = true;
      this.fieldName = getFieldName(parentGroupName, this.props.name).join('-');
      this.collect = collect !== false;
      this.source = initialSource;
      this.visible = initialVisible;
      this.disabled = initialDisabled;
    }

    const formItemProps = restProps as FormItemProps;

    if (restProps.rules && restProps.rules.length) {
      formItemProps.rules = restProps.rules.map((rule) => {
        if (rule.validator) {
          return {
            validator: (_, value): Promise<void> | void => {
              const validatorCtx = getCtx({
                value,
                name: this.props.name,
              });
              return triggerPlugin('validator', validatorCtx, rule.validator);
            },
          };
        }
        return rule;
      });
    }

    const { getCtx } = formContext.getInternalHooks(HOOK_MARK);
    const ctx = getCtx({ disabled: this.disabled, source: this.source, name: this.props.name });

    const extraConfig = {
      layout,
      itemLayout: isList
        ? itemLayout || {}
        : assignWithDefined({}, this.context.itemLayout, itemLayout),
      visible: this.visible,
      childrenContainer,
      formItemProps,
      isList,
    };

    const fieldNode = this.renderField(ctx, fieldProps, extraConfig);

    const subFormContext = {
      formContext,
      parentGroupName,
      layout,
    };

    return (
      <FieldGroupContext.Provider value={subFormContext}>{fieldNode}</FieldGroupContext.Provider>
    );
  }
}

function getFieldName(parentName: string, name?: string | string[]) {
  const finalName = name ? toArray(name) : [uniqueId('fieldName_')];
  return [parentName, ...finalName];
}
