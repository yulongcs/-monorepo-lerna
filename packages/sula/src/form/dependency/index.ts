import isEmpty from 'lodash/isEmpty';
import omit from 'lodash/omit';
import { FieldNameList, FieldNamePath } from '../../types/form';
import { FieldProps } from '../Field';
import { HOOK_MARK } from '../FieldGroupContext';
import DepStore from './DepStore';

const DEP_NAME = 'dependency';

export default class FormDependency {
  private depStore: DepStore;

  constructor() {
    this.depStore = new DepStore();
  }

  public parseFormDependency = (
    fieldConfig: FieldProps,
    getName: (name: FieldNamePath) => FieldNameList,
  ) => {
    const dependencies = fieldConfig[DEP_NAME];
    if (isEmpty(dependencies)) {
      return fieldConfig;
    }
    this.depStore.parse(fieldConfig, dependencies!, getName);
    return omit(fieldConfig, [DEP_NAME]);
  };

  public removeDependency = (fieldNameList) => {
    this.depStore.depsByFieldNameList.delete(fieldNameList);
  }

  public getCascades = (): FieldNameList[] => this.depStore.depsByFieldNameList.getNameLists();

  public cascade = (ctx, store, cascadePayload) => {
    const { getFieldKeyByFieldName } = ctx.form.getInternalHooks(HOOK_MARK);
    store.forEach((item) => {
      const cascadeFieldNameList = getFieldKeyByFieldName(item.name);
      const depsOfType = this.depStore.depsByFieldNameList.get(cascadeFieldNameList);
      if (depsOfType) {
        this.depStore.triggerDependency(ctx, depsOfType, cascadePayload);
      }
    });
  };
}
