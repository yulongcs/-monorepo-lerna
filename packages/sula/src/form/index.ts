import React from 'react';
import Field from './Field';
import FormList from './FormList';
import FieldGroup from './FieldGroup';
import FormAction from './FormAction';
import MediaQueries from './MediaQueries';
import {
  default as FieldForm,
  FormProps as InternalFormProps,
  FormInstance as FormInstanceProps,
} from './Form';
import { FieldGroupProps as InternalFieldGroupProps } from './FieldGroup';
import { FieldProps as InternalFieldProps } from './Field';
import useFormContext from './useFormContext';

const InternalForm = React.forwardRef<FormInstanceProps, InternalFormProps>(FieldForm) as (
  props: React.PropsWithChildren<InternalFormProps> & { ref?: React.Ref<FormInstanceProps> },
) => React.ReactElement;

type InternalForm = typeof InternalForm;

const useForm = useFormContext;
interface RefForm extends InternalForm {
  Field: typeof Field;
  FieldGroup: typeof FieldGroup;
  FormList: typeof FormList;
  useForm: typeof useForm;
}

const RefForm: RefForm = InternalForm as RefForm;

RefForm.Field = Field;
RefForm.FieldGroup = FieldGroup;
RefForm.FormList = FormList;
RefForm.useForm = useForm;

const Form = RefForm;

export { Form, Field, FieldGroup, FormList, FormAction, MediaQueries, useForm };

export interface FormProps extends InternalFormProps {}
export interface FieldGroupProps extends InternalFieldGroupProps {}
export interface FieldProps extends InternalFieldProps {}
export interface FormInstance extends FormInstanceProps {}
