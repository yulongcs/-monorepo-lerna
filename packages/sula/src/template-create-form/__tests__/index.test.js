import React from 'react';
import { mount } from 'enzyme';
import { CreateForm } from '..';
import { Form } from '../../form';
import { delay } from '../../__tests__/common';

const fields = [
  {
    name: 'input1',
    label: 'label1',
    field: 'input',
  },
  {
    name: 'input2',
    label: 'label2',
    field: 'input',
  },
  {
    name: 'input3',
    label: 'label3',
    field: 'input',
  },
];

const submit = {
  url: '/submit.json',
  method: 'post',
};

describe('createForm', () => {
  it('createForm view', () => {
    const wrapper = mount(<CreateForm fields={fields} mode="view" />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('createForm props', () => {
    const wrapper = mount(
      <CreateForm
        fields={fields}
        actionsPosition="bottom"
        submitButtonProps={{
          children: 'SubmitTest',
        }}
        backButtonProps={{
          children: 'CancelTest',
          disabled: true,
        }}
        submit={submit}
      />,
    );
    expect(wrapper.find('.sula-form-action-bottom').length).toEqual(1);
    expect(wrapper.find('button').first().text()).toEqual('SubmitTest');
    expect(wrapper.find('button').at(1).text()).toEqual('CancelTest');
    expect(wrapper.find('button').at(1).props().disabled).toEqual(true);
  });

  it('createForm back', () => {
    const fn = jest.fn();
    const wrapper = mount(<CreateForm fields={fields} back={fn} submit={submit} />);
    wrapper.find('button').at(1).simulate('click');
    expect(fn).toHaveBeenCalled();
  });

  it('createForm itemLayout cols', () => {
    const wrapper = mount(
      <CreateForm
        fields={fields}
        itemLayout={{
          cols: 1,
        }}
        submit={submit}
      />,
    );
    expect(wrapper.find(Form).props().actionsPosition).toEqual('default');

    wrapper.setProps({ itemLayout: { cols: 3 } });
    expect(wrapper.find(Form).props().actionsPosition).toEqual('center');
  });

  it('createForm itemLayout cols auto', () => {
    const wrapper = mount(
      <CreateForm
        itemLayout={{
          cols: {
            xl: 3,
            sm: 2,
          },
        }}
        fields={fields}
        submit={submit}
      />,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('createForm actionsRender', () => {
    const wrapper = mount(
      <CreateForm
        fields={fields}
        actionsRender={[
          {
            type: 'button',
            props: {
              children: 'button',
            },
            action: jest.fn(),
          },
        ]}
      />,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('loading', async () => {
    const remoteValues = async () => {
      await delay(2000);
      return Promise.resolve({
        input1: 'a',
        input2: 'b',
      });
    };
    const wrapper = mount(
      <CreateForm fields={fields} submit={submit} mode="edit" remoteValues={remoteValues} />,
    );
    await delay(1000);
    wrapper.update();
    expect(wrapper.find('Spin').props().spinning).toEqual(true);

    await delay(2500);
    wrapper.update();
    expect(wrapper.find('Spin').props().spinning).toEqual(false);
  });

  it('createForm submit', async (done) => {
    const initialValues = {
      input1: 'input1',
      input2: 'input2',
      input3: 'input3',
    };
    const wrapper = mount(
      <CreateForm
        fields={fields}
        initialValues={initialValues}
        submit={{
          ...submit,
          converter: ({ data }) => {
            expect(data).toEqual(initialValues);
            done();
          },
        }}
      />,
    );
    await wrapper.find('button').first().simulate('click');
  });
});
