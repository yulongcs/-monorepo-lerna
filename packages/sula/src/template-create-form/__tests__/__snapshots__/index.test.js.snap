// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createForm createForm actionsRender 1`] = `
<div
  class="ant-spin-nested-loading"
>
  <div
    class="ant-spin-container"
  >
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-col-8 ant-form-item-label"
        >
          <label
            class=""
            for="input1"
            title="label1"
          >
            label1
          </label>
        </div>
        <div
          class="ant-col ant-col-12 ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="input1"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-col-8 ant-form-item-label"
        >
          <label
            class=""
            for="input2"
            title="label2"
          >
            label2
          </label>
        </div>
        <div
          class="ant-col ant-col-12 ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="input2"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-col-8 ant-form-item-label"
        >
          <label
            class=""
            for="input3"
            title="label3"
          >
            label3
          </label>
        </div>
        <div
          class="ant-col ant-col-12 ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="input3"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
        style="margin-bottom: 0px;"
      >
        <div
          class="ant-col ant-col-12 ant-col-offset-8 ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="sula-form-action-default"
              >
                <button
                  class="ant-btn"
                  type="button"
                >
                  <span>
                    button
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`createForm createForm itemLayout cols auto 1`] = `
<div
  class="ant-spin-nested-loading"
>
  <div
    class="ant-spin-container"
  >
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-row"
        style="margin-left: -12px; margin-right: -12px;"
      >
        <div
          class="ant-col ant-col-NaN"
          style="padding-left: 12px; padding-right: 12px;"
        >
          <div
            class="ant-row ant-form-item"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class=""
                for="input1"
                title="label1"
              >
                label1
              </label>
            </div>
            <div
              class="ant-col ant-col-16 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input"
                    id="input1"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-NaN"
          style="padding-left: 12px; padding-right: 12px;"
        >
          <div
            class="ant-row ant-form-item"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class=""
                for="input2"
                title="label2"
              >
                label2
              </label>
            </div>
            <div
              class="ant-col ant-col-16 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input"
                    id="input2"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-NaN"
          style="padding-left: 12px; padding-right: 12px;"
        >
          <div
            class="ant-row ant-form-item"
          >
            <div
              class="ant-col ant-col-8 ant-form-item-label"
            >
              <label
                class=""
                for="input3"
                title="label3"
              >
                label3
              </label>
            </div>
            <div
              class="ant-col ant-col-16 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    class="ant-input"
                    id="input3"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-NaN"
          style="padding-left: 12px; padding-right: 12px;"
        >
          <div
            class="ant-row ant-form-item"
            style="margin-bottom: 0px;"
          >
            <div
              class="ant-col ant-col-16 ant-col-offset-8 ant-form-item-control"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <div
                    class="sula-form-action-default"
                  >
                    <div
                      class="ant-space ant-space-horizontal ant-space-align-center"
                    >
                      <div
                        class="ant-space-item"
                        style="margin-right: 8px;"
                      >
                        <button
                          class="ant-btn ant-btn-primary"
                          type="button"
                        >
                          <span>
                            Submit
                          </span>
                        </button>
                      </div>
                      <div
                        class="ant-space-item"
                      >
                        <button
                          class="ant-btn"
                          type="button"
                        >
                          <span>
                            Cancel
                          </span>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`createForm createForm view 1`] = `
<div
  class="ant-spin-nested-loading"
>
  <div
    class="ant-spin-container"
  >
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-col-8 ant-form-item-label"
        >
          <label
            class=""
            for="input1"
            title="label1"
          >
            label1
          </label>
        </div>
        <div
          class="ant-col ant-col-12 ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input ant-input-disabled"
                disabled=""
                id="input1"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-col-8 ant-form-item-label"
        >
          <label
            class=""
            for="input2"
            title="label2"
          >
            label2
          </label>
        </div>
        <div
          class="ant-col ant-col-12 ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input ant-input-disabled"
                disabled=""
                id="input2"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-col-8 ant-form-item-label"
        >
          <label
            class=""
            for="input3"
            title="label3"
          >
            label3
          </label>
        </div>
        <div
          class="ant-col ant-col-12 ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input ant-input-disabled"
                disabled=""
                id="input3"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
        style="margin-bottom: 0px;"
      >
        <div
          class="ant-col ant-col-12 ant-col-offset-8 ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="sula-form-action-default"
              >
                <button
                  class="ant-btn"
                  type="button"
                >
                  <span>
                    Back
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
